containers/payments/subscription/NewSubscriptionModal.js

```diff
@@ -0,0 +1,357 @@
+                    Cycle: model.cycle,
+                    ...params
+                })
+            );
+            await call();
+            setStep(STEPS.THANKS);
+        } catch (error) {
+            setStep(STEPS.PAYMENT);
+            throw error;
+        }
+    };
+
+    const { card, setCard, errors, method, setMethod, parameters, canPay, paypal, paypalCredit } = usePayment({
+        amount: checkResult.AmountDue,
+        currency: checkResult.Currency,
+        onPay(params) {
+            return withLoading(handleSubscribe(params));
+        }
+    });
+
+    const SubmitButton = ({ className }) => {
+        if (step === STEPS.CUSTOMIZATION) {
+            return (
+                <PrimaryButton className={className} loading={loadingCheck} type="submit">{c('Action')
+                    .t`Continue`}</PrimaryButton>
+            );
+        }
+
+        if (method === PAYMENT_METHOD_TYPES.PAYPAL) {
+            return (
+                <PayPalButton paypal={paypal} className={className} amount={checkResult.AmountDue}>{c('Action')
+                    .t`Pay`}</PayPalButton>
+            );
+        }
+
+        if ([PAYMENT_METHOD_TYPES.CASH, PAYMENT_METHOD_TYPES.BITCOIN].includes(method)) {
+            return (
+                <PrimaryButton className={className} loading={loadingCheck} onClick={() => setStep(STEPS.THANKS)}>{c(
+                    'Action'
+                ).t`Done`}</PrimaryButton>
+            );
+        }
+
+        if (!checkResult.AmountDue) {
+            return (
+                <PrimaryButton className={className} loading={loadingCheck} disabled={!canPay} type="submit">{c(
+                    'Action'
+                ).t`Complete`}</PrimaryButton>
+            );
+        }
+
+        return (
+            <PrimaryButton className={className} loading={loadingCheck} disabled={!canPay} type="submit">{c('Action')
+                .t`Pay`}</PrimaryButton>
+        );
+    };
+
+    SubmitButton.propTypes = {
+        className: PropTypes.string
+    };
+
+    const check = async (newModel = model) => {
+        if (!hasPlans(newModel.planIDs)) {
+            setCheckResult(TOTAL_ZERO);
+            return;
+        }
+
+        try {
+            const result = await api(
+                checkSubscription({
+                    PlanIDs: clearPlanIDs(newModel.planIDs),
+                    CouponCode: newModel.coupon,
+                    Currency: newModel.currency,
+                    Cycle: newModel.cycle,
+                    GiftCode: newModel.gift
+                })
+            );
+
+            const { Code = '' } = result.Coupon || {}; // Coupon can equal null
+            newModel.coupon = Code;
+
+            if (!result.Gift) {
+                delete newModel.gift;
+            }
+
+            setModel(newModel);
+            setCheckResult(result);
+        } catch (error) {
+            setModel(model);
+            throw error;
+        }
+    };
```

Comment:
I find it a bit confusing that we have to `setModel(model)` if there was an error, since presumably any error should happen before the setState's of the `try`, making redundant to set the model to its current value. Also, the `delete newModel.gift` is altering the argument of the function `check`, which I think in general it's something we try to avoid. Couldn't we just create a copy of `newModel` inside the function and alter that one?